### PR TITLE
feature: uri pattern for test mocking

### PR DESCRIPTION
**Pull request by MidnightWonderer**
Add regular expression to middleware matcher.

### CODE REVIEW

**Comment by MidnightWonderer** [comment]
The failed test has nothing to do with the feature in this pull request.
```
Gem::InstallError: addressable requires Ruby version >= 1.9.0.
Installing rack-test 0.6.3
An error occurred while installing addressable (2.4.0), and Bundler cannot
```

File: lib/faraday/adapter/test.rb [hunk 1]
```diff
@@ -85,7 +85,7 @@ def verify_stubbed_calls
         protected
 
         def new_stub(request_method, path, headers = {}, body=nil, &block)
-          normalized_path = Faraday::Utils.normalize_path(path)
+          normalized_path = (path.is_a? Regexp) ? path : Faraday::Utils.normalize_path(path)
```
Comment (mislav) [inline lib/faraday/adapter/test.rb]: Style nitpick: could it be `path.is_a?(Regexp)` instead?

File: lib/faraday/adapter/test.rb [hunk 2]
```diff
@@ -96,7 +96,7 @@ def matches?(stack, path, headers, body)
 
       class Stub < Struct.new(:path, :params, :headers, :body, :block)
         def initialize(full, headers, body, block)
-          path, query = full.split('?')
+          path, query = (full.is_a? Regexp) ? full : full.split('?')
```
Comment (mislav) [inline lib/faraday/adapter/test.rb]: I feel like it's better to use `full.respond_to?(:split) ? full.split("?") ...`

File: lib/faraday/adapter/test.rb [hunk 3]
```diff
@@ -108,12 +108,20 @@ def matches?(request_uri, request_headers, request_body)
...
-          request_path == path &&
+          path_match?(request_path) &&
             params_match?(request_params) &&
             (body.to_s.size.zero? || request_body == body) &&
             headers_match?(request_headers)
         end
 
+        def path_match?(request_path)
+          if path.is_a? Regexp
+            path =~ request_path
```
Comment (mislav) [inline lib/faraday/adapter/test.rb]: Could we capture the MatchData object by doing `match = request_path.match(path)` and somehow pass it to the block? That way the block could access the values from capture groups

**Comment by mislav** [comment]
Thanks for the feature proposal! Sorry about the CI failure; you can merge in latest master to fix that.

In addition to addressing my thoughts in inline comments, could you also add code docs for this feature? I realize there is very little documentation for the Test adapter right now. Your PR could try to improve this while adding a new feature. At least some example use would be nice.

**Comment by MidnightWonderer** [comment]
@mislav all done
Let me know if you have any more suggestion for new code added.

**Comment by iMacTia** [comment]
Hi @MidnightWonderer and sorry if your PR remained suspended for so much time.
I would like to merge it in, but I think that tests are outdated now. Can you please pull latest changes from master once more? thanks in advance

**Comment by MidnightWonderer** [comment]
@iMacTia done.
The tests are still passed without doing anything.

**Comment by iMacTia** [comment]
This took a long time, but I'm happy to merge it in as it actually makes tests more powerful.
Good work @MidnightWonderer 👍 !

**Pull request merged**